### DISCUSSION TITLE
Real games using ursina

### DISCUSSION THREAD

**Discussion by rbignon** [pokepetter/ursina#699] [Q&A]
Hello,

I want to write a game using ursina. However, the only source code I found on github are little games or examples, but not real games.

There are two questions here:

* do you know a repository where I can read source code of an advanced game?
* for me, this is a problem that `update` and `input` functions have to be global, as I want to write my code using classes. I read `main.py` and it seems there is no alternative. Maybe I should override the `Ursina` class instead? What is the recommandation when you want to write a game application which is not a single file script?

Thank you.

**Comment by pokepetter** [comment]
Global `update` and `input` is not necessary. It was intended as a convenience feature when writing small single file scripts. The best way is to make and Entity (inherit) and implement update and input for that. This makes it possible to enabled/disable the logic at will, and to decide which order things should run in if you have multiple entities. 
See: https://www.ursinaengine.org/entity_basics.html#Update

You should not override `Ursina`. It's currently not even possible. It was easy to override existing functions this way, like input, which would break input should you forget to call the parent method.

**Comment by rbignon** [comment]
Thank you for your quick answer.

Ok got it, it makes sense. Do you know an example project which uses Entity's methods inheritance instead of global functions?

**Comment by a2cy** [comment]
I don't know if it helps but I've made this https://github.com/a2cy/CubePixel/tree/upstream

**Reply by rbignon** [reply]
Hello,

Thank you for this link. I tried to use it, but there is this crash:

```
$ python main.py
info: No primary monitor found, using first monitor instead
:prc(warning): Invalid integer value for ConfigVariable win-size: 1152.0
:prc(warning): Invalid integer value for ConfigVariable win-size: 2752.0
Known pipe types:
  glxGraphicsPipe
(4 aux display modules not yet loaded.)
set window position: Vec2(860, 18.5)
:pnmimage:png(warning): iCCP: known incorrect sRGB profile
package_folder: /home/rom1/tmp/CubePixel/venv/lib/python3.12/site-packages/ursina
asset_folder: .
Traceback (most recent call last):
  File "/home/rom1/tmp/CubePixel/CubePixel/main.py", line 105, in <module>
:thread(error): Exception occurred within Thread TaskManager_chunk_update_0
Traceback (most recent call last):
  File "/home/rom1/tmp/CubePixel/CubePixel/modules/chunk_handler.py", line 329, in _update
    main()
  File "/home/rom1/tmp/CubePixel/CubePixel/main.py", line 99, in main
    CubePixel(app=app)
  File "/home/rom1/tmp/CubePixel/venv/lib/python3.12/site-packages/ursina/ursinastuff.py", line 238, in __call__
    if not self.world_loaded:
           ^^^^^^^^^^^^^^^^^
    obj = type.__call__(cls, *args, **kwargs)
AttributeError: 'ChunkHandler' object has no attribute 'world_loaded'
:task(error): Exception occurred in PythonTask chunk_update
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rom1/tmp/CubePixel/CubePixel/main.py", line 39, in __init__
    self.chunk_handler = ChunkHandler(self)
                         ^^^^^^^^^^^^^^^^^^
  File "/home/rom1/tmp/CubePixel/venv/lib/python3.12/site-packages/ursina/ursinastuff.py", line 238, in __call__
    obj = type.__call__(cls, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rom1/tmp/CubePixel/CubePixel/modules/chunk_handler.py", line 29, in __init__
    self.render_distance = self.game.settings["render_distance"]
    ^^^^^^^^^^^^^^^^^^^^
  File "/home/rom1/tmp/CubePixel/CubePixel/modules/chunk_handler.py", line 58, in render_distance
    if self.world_loaded:
       ^^^^^^^^^^^^^^^^^
AttributeError: 'ChunkHandler' object has no attribute 'world_loaded'
```

(I'm in a venv where I correctly installed requirements.txt and built ext, I don't know if this is a problem with the version of ursina or something)

**Reply by a2cy** [reply]
you used the wrong branch :) the main branch is currently broken

**Reply by rbignon** [reply]
Ah ok, my bad, thx :)
